app/waf/test/acceptance/application.controller.acceptance.ts

```diff
@@ -264,4 +231,104 @@ describe('ApplicationController', () => {
       .send(application)
       .expect(404);
   });
+
+  it(
+    'post ' +
+      prefix +
+      '/applications/{id}/deploy: no ADC associated with application',
+    async () => {
+      const application = await givenApplicationData(wafapp);
+
+      await client
+        .post(prefix + '/applications/' + application.id + '/deploy')
+        .expect(422);
+    },
+  );
+
+  it(
+    'post ' + prefix + '/applications/{id}/deploy: application has no service',
+    async () => {
+      const application = await givenApplicationData(wafapp);
+      const adc = await givenAdcData(wafapp);
+      await givenTenantAssociationData(wafapp, {
+        tenantId: 'default',
+        adcId: adc.id,
+      });
+
+      await client
+        .post(prefix + '/applications/' + application.id + '/deploy')
+        .expect(422);
+    },
+  );
+
+  it(
+    'post ' + prefix + '/applications/{id}/deploy: deploy as3 config',
+    async () => {
+      const adc = await givenAdcData(wafapp);
+      await givenTenantAssociationData(wafapp, {
+        tenantId: 'default',
+        adcId: adc.id,
+      });
+      let member = await givenMemberData(wafapp);
+      let pool = await givenPoolData(wafapp, {
+        name: 'pool1',
+        members: [member.id],
+      });
+      let service = await givenServiceData(wafapp, {
+        pool: pool.id,
+      });
+      let application = await givenApplicationData(wafapp, {
+        services: [service.id],
+      });
+
+      let s = sinon
+        .stub(controller.as3Service, 'deploy')
```

Comment:
How about to think about how to abstract the stub framework for rest datasource if we prefer to stub it.

Reply:
I think it is already a 'ready-to-use' framework. What is your expectation?

Reply:
Agree, restServer for restDatasource is not required. 

Here, I just want to make the `.stub(controller.as3Service, 'deploy')` reusable, which means move the stub statement to `before` and `after`.

Reply:
These two parameters make stub() to be reusable. Every test case has its unique stub requirement, eg. difference class, function, stub behaviors. It does not make sense to move it to before().

Reply:
NO. 

You have several it() has the stub, they are duplicate code, encapsulation should be done. 
Move them to `before` and `after`.  Make the code as clear as possible.

Reply:
You are writing acceptance test. not unit test.
"
difference class, function, stub behaviors. " should not be the excuse of not doing encapsulation and abstraction.